cmd/update/volume_move.go

```diff
@@ -0,0 +1,383 @@
+	useIDs, err := c.config.UseIDs()
+	if err != nil {
+		return err
+	}
+
+	// fetch IDs
+	nsID := id.Namespace(c.namespace)
+
+	if !useIDs {
+		ns, err := c.client.GetNamespaceByName(ctx, c.namespace)
+		if err != nil {
+			return err
+		}
+		nsID = ns.ID
+
+		fromNode, err := c.client.GetNodeByName(ctx, c.fromNode)
+		if err != nil {
+			return err
+		}
+		c.fromNode = fromNode.ID.String()
+
+		toNode, err := c.client.GetNodeByName(ctx, c.toNode)
+		if err != nil {
+			return err
+		}
+		c.toNode = toNode.ID.String()
+	}
+
+	// fetch volume
+	vol, err := c.getVolume(ctx, nsID)
+	if err != nil {
+		return err
+	}
+
+	if vol.Master.Health != health.MasterOnline {
+		return fmt.Errorf("cannot move volume - master unhealthy: %v", vol.Master.Health.String())
+	}
+	for _, replica := range vol.Replicas {
+		if replica.Health != health.ReplicaReady || !replica.Promotable {
+			return fmt.Errorf("cannot move volume - replica unhealthy, id: %v health: %v, promotable: %v", replica.ID, replica.Health, replica.Promotable)
+		}
+	}
+
+	hasReplicaOnSrcNode := false
+	isRetryRun := false
+	var deploymentOnSourceNode id.Deployment
+	for _, rep := range vol.Replicas {
+		// make sure there's at least one replica on the [source] node
+		if rep.Node.String() == c.fromNode {
+			hasReplicaOnSrcNode = true
+			deploymentOnSourceNode = rep.ID
+		}
+
+		// if there's already a replica on the [destination] node we will assume
+		// we are retrying a previously failed request (timed out would be the
+		// most likely failure)
+		if rep.Node.String() == c.toNode {
+			isRetryRun = true
+		}
+	}
+
+	isMovingMaster := false
+	if vol.Master.Node.String() == c.fromNode {
+		isMovingMaster = true
+	} else if !hasReplicaOnSrcNode {
+		return fmt.Errorf("volume has no replicas on [from] node %s", c.fromNode)
+	}
+
+	// create the new replica on [destination] node
+	// skip if there's one one already (retry run)
+	if !isRetryRun {
+		params := &apiclient.AddDeploymentOnNodeRequestParams{
+			NodeID: id.Node(c.toNode),
+		}
+
+		if c.useCAS() {
+			params.CASVersion = version.FromString(c.casVersion)
+		}
+
+		err = c.client.AddDeploymentOnNode(ctx, nsID, vol.ID, params)
+		if err != nil {
+			return fmt.Errorf("failed to add deployment on node %s: %s", c.toNode, err)
+		}
+	}
+
+	// wait for a new replica on [destination] node
+	err = c.waitForReplicaOnNode(ctx, nsID)
+	if err != nil {
+		return err
+	}
+
+	// get updated volume config
+	vol, err = c.getVolume(ctx, nsID)
+	if err != nil {
+		return fmt.Errorf("failed to get updated volume config from api: %s", err)
+	}
+
+	if !isMovingMaster {
+		// mark the replica on the [source] node as preferred for eviction
+		//
+		// the master node needs to first be moved and then be evicted due to
+		// api endpoints restriction
+		params := &apiclient.SetPreferredEvictionCandidatesRequestParams{
+			DeploymentIDs: []string{deploymentOnSourceNode.String()},
+		}
+		if c.useCAS() {
+			params.CASVersion = vol.Version
```

Comment:
At this point our CAS might be outdated if have ran `err = c.client.AddDeploymentOnNode(ctx, nsID, vol.ID, params)`, so we should probably just grab the latest from the volume

Reply:
We are getting the updated volume config just above this if statement line ~138
Is that not what you mean here?

Reply:
You're right. I misread this